File: fetch_amazon_data.js
const fs = require('fs');
const path = require('path');

// --- CONFIGURATION ---
const API_KEY = null; // Set this to your actual key to fetch live data (e.g., Rainforest API)

async function fetchRainforestData(keyword, apiKey) {
    if (!apiKey) {
        console.log("⚠️  No API Key provided. Switching to MOCK DATA mode.");
        return null;
    }
    // Simulation of API call
    console.log(`Fetching live data for ${keyword}... (API Key logic would go here)`);
    return null;
}

function estimateSalesFromBsr(bsr, category = "Beauty") {
    // Simplified heuristic logic
    if (!bsr) return 0;

    if (bsr < 10) return 50000;
    else if (bsr < 100) return 10000;
    else if (bsr < 1000) return 3000;
    else if (bsr < 5000) return 1000;
    else if (bsr < 10000) return 300;
    else if (bsr < 50000) return 50;
    else return 10;
}

async function analyzeMarket(keyword) {
    console.log(`\n🔍 Analyzing Market for Keyword: '${keyword}'...`);
    console.log("-".repeat(50));

    let data = await fetchRainforestData(keyword, API_KEY);

    if (!data) {
        try {
            const dataPath = path.join(__dirname, 'sample_data.json');
            const fileContent = fs.readFileSync(dataPath, 'utf8');
            data = JSON.parse(fileContent);
            console.log("✅ Loaded SAMPLE DATA (simulating API response)");
        } catch (err) {
            console.error("❌ Error: Sample data file not found.", err.message);
            return;
        }
    }

    const results = data.search_results || [];
    if (results.length === 0) {
        console.log("No results found.");
        return;
    }

    // Extract Metrics
    const prices = [];
    const competitors = [];

    results.forEach(item => {
        const priceObj = item.price;
        const price = priceObj ? priceObj.value : null;
        const bsr = item.best_seller_rank;
        const title = item.title || 'Unknown';

        const salesEst = estimateSalesFromBsr(bsr);

        if (price !== null) {
            prices.push(price);
        }

        competitors.push({
            title: title.length > 50 ? title.substring(0, 50) + "..." : title,
            price: price,
            bsr: bsr,
            est_sales: salesEst,
            revenue: (price || 0) * salesEst
        });
    });

    if (prices.length === 0) {
        console.log("No pricing data available.");
        return;
    }

    const avgPrice = prices.reduce((a, b) => a + b, 0) / prices.length;
    const minPrice = Math.min(...prices);
    const maxPrice = Math.max(...prices);
    const totalEstSales = competitors.reduce((sum, item) => sum + item.est_sales, 0);
    const estMarketRevenue = competitors.reduce((sum, item) => sum + item.revenue, 0);

    // Generate Report
    console.log(`\n📊 MARKET INTELLIGENCE REPORT: ${keyword}`);
    console.log("=".repeat(50));
    console.log(`💰 Average Price:       $${avgPrice.toFixed(2)}`);
    console.log(`📉 Price Range:         $${minPrice.toFixed(2)} - $${maxPrice.toFixed(2)}`);
    console.log(`📦 Est. Monthly Sales (Top ${results.length}): ${totalEstSales.toLocaleString()} units`);
    console.log(`💵 Est. Monthly Revenue (Top ${results.length}): $${estMarketRevenue.toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`);
    console.log("-".repeat(50));
    console.log("🏆 TOP COMPETITORS:");

    competitors
        .sort((a, b) => b.revenue - a.revenue)
        .slice(0, 5)
        .forEach((product, i) => {
            console.log(`${i + 1}. ${product.title}`);
            console.log(`   Price: $${product.price} | BSR: #${product.bsr} | Est. Sales: ~${product.est_sales}/mo`);
            console.log(`   Est. Rev: $${product.revenue.toLocaleString()}`);
            console.log("");
        });

    console.log("💡 STRATEGIC RECOMMENDATION:");
    if (avgPrice > 20 && totalEstSales > 5000) {
        console.log("   ✅ GO: High demand and healthy price margins detected.");
        console.log("   Strategy: Bundle products to increase Average Order Value (AOV) above $25.");
    } else if (avgPrice < 15) {
        console.log("   ⚠️ CAUTION: Price war zone. Low margins expected.");
        console.log("   Strategy: Focus on extreme differentiation or branding to command higher prices.");
    } else {
        console.log("   ℹ️ NEUTRAL: Standard market conditions. Requires good marketing.");
    }
}

analyzeMarket("Korean Skincare");

File: fetch_amazon_data.py
import json
import os
import requests
import statistics

# --- CONFIGURATION ---
# In a real scenario, you would get an API Key from a provider like Rainforest API, DataForSEO, or RapidAPI.
# provider_url = "https://api.rainforestapi.com/request"
API_KEY = None # Set this to your actual key to fetch live data

def fetch_rainforest_data(keyword, api_key):
    """
    Fetches live data from Rainforest API (Simulated implementation).
    """
    if not api_key:
        print("⚠️  No API Key provided. Switching to MOCK DATA mode.")
        return None

    # This is how the real request would look:
    # params = {
    #   'api_key': api_key,
    #   'type': 'search',
    #   'amazon_domain': 'amazon.com',
    #   'search_term': keyword,
    #   'sort_by': 'most_recent'
    # }
    # response = requests.get('https://api.rainforestapi.com/request', params=params)
    # return response.json()
    return None

def estimate_sales_from_bsr(bsr, category="Beauty"):
    """
    Roughly estimates monthly sales based on Best Seller Rank (BSR).
    NOTE: This is a simplified heuristic. Real tools use massive databases to define these curves.
    Curve assumption for Beauty category: very high sales at top ranks, dropping logarithmically.
    """
    if not bsr:
        return 0
    
    # Simple logarithmic decay model query for demonstration
    # Rank 1 ~ 30,000 sales
    # Rank 100 ~ 6,000 sales 
    # Rank 1,000 ~ 900 sales
    # Rank 10,000 ~ 100 sales
    
    if bsr < 10:
        return 50000
    elif bsr < 100:
        return 10000
    elif bsr < 1000:
        return 3000
    elif bsr < 5000:
        return 1000
    elif bsr < 10000:
        return 300
    elif bsr < 50000:
        return 50
    else:
        return 10

def analyze_market(keyword):
    print(f"\n🔍 Analyzing Market for Keyword: '{keyword}'...")
    print("-" * 50)

    # 1. Load Data (Mock or Real)
    data = fetch_rainforest_data(keyword, API_KEY)
    
    if not data:
        # Load local sample data for demonstration
        try:
            with open('sample_data.json', 'r', encoding='utf-8') as f:
                data = json.load(f)
            print("✅ Loaded SAMPLE DATA (simulating API response)")
        except FileNotFoundError:
            print("❌ Error: Sample data file not found.")
            return

    results = data.get('search_results', [])
    if not results:
        print("No results found.")
        return

    # 2. Extract Metrics
    prices = []
    sales_estimates = []
    competitors = []

    for item in results:
        # Price extraction
        price_obj = item.get('price')
        price = price_obj.get('value') if price_obj else None
        
        # BSR extraction
        bsr = item.get('best_seller_rank')
        
        # Title extraction
        title = item.get('title', 'Unknown')

        if price:
            prices.append(price)
            
        sales_est = estimate_sales_from_bsr(bsr)
        sales_estimates.append(sales_est)
        
        competitors.append({
            'title': title[:50] + "...",
            'price': price,
            'bsr': bsr,
            'est_sales': sales_est,
            'revenue': price * sales_est if price else 0
        })

    # 3. Calculate Aggregates
    if not prices:
        print("No pricing data available.")
        return

    avg_price = statistics.mean(prices)
    min_price = min(prices)
    max_price = max(prices)
    total_est_monthly_sales = sum(sales_estimates)
    avg_est_monthly_sales = statistics.mean(sales_estimates)
    est_market_revenue = sum(c['revenue'] for c in competitors)

    # 4. Generate Report
    print(f"\n📊 MARKET INTELLIGENCE REPORT: {keyword}")
    print("=" * 50)
    print(f"💰 Average Price:       ${avg_price:.2f}")
    print(f"📉 Price Range:         ${min_price:.2f} - ${max_price:.2f}")
    print(f"📦 Est. Monthly Sales (Top {len(results)}): {total_est_monthly_sales:,} units")
    print(f"💵 Est. Monthly Revenue (Top {len(results)}): ${est_market_revenue:,.2f}")
    print("-" * 50)
    print("🏆 TOP COMPETITORS:")
    
    # Sort by estimated revenue
    sorted_competitors = sorted(competitors, key=lambda x: x['revenue'], reverse=True)
    
    for i, product in enumerate(sorted_competitors[:5], 1):
        print(f"{i}. {product['title']}")
        print(f"   Price: ${product['price']} | BSR: #{product['bsr']} | Est. Sales: ~{product['est_sales']}/mo")
        print(f"   Est. Rev: ${product['revenue']:,.0f}")
        print("")

    # 5. Strategic Advice (Simple Logic)
    print("💡 STRATEGIC RECOMMENDATION:")
    if avg_price > 20 and total_est_monthly_sales > 5000:
        print("   ✅ GO: High demand and healthy price margins detected.")
        print("   Strategy: Bundle products to increase Average Order Value (AOV) above $25.")
    elif avg_price < 15:
        print("   ⚠️ CAUTION: Price war zone. Low margins expected.")
        print("   Strategy: Focus on extreme differentiation or branding to command higher prices.")
    else:
        print("   ℹ️ NEUTRAL: Standard market conditions. Requires good marketing.")

if __name__ == "__main__":
    analyze_market("Korean Skincare")
